File: leetcode/101-200/172-factorial-trailing-zeroes/trailing_zeroes.py
"""LeetCode 172 — Factorial Trailing Zeroes (Python mirror / oracle).

Count factors of 5 in n!: sum ⌊n/5⌋ + ⌊n/25⌋ + … by repeatedly dividing (every
intermediate stays below n, so no overflow). Mirrors the Kāra version.
"""


def trailing_zeroes(n):
    count = 0
    m = n // 5
    while m > 0:
        count += m
        m //= 5
    return count


def report(n):
    print(trailing_zeroes(n))


def main():
    report(3)
    report(5)
    report(0)
    report(10)
    report(25)
    report(100)
    report(1000)
    report(10000)


main()

File: leetcode/101-200/172-factorial-trailing-zeroes/bench/trailing_zeroes.c
#include <stdio.h>

static long trailing_zeroes(long n) {
    long count = 0;
    long m = n / 5;
    while (m > 0) {
        count += m;
        m /= 5;
    }
    return count;
}

int main(void) {
    long limit = 35000000;
    long sink = 0;
    for (long i = 0; i < limit; i++) {
        sink += trailing_zeroes(i);
    }
    printf("%ld\n", sink);
    return 0;
}
